core/src/main/java/org/testcontainers/containers/image/ImageNameResolverProvider.java

```diff
@@ -0,0 +1,39 @@
+package org.testcontainers.containers.image;
+
+import java.util.concurrent.Future;
+import org.testcontainers.containers.image.pull.policy.ImagePullPolicy;
+import org.testcontainers.containers.image.pull.policy.PullPolicy;
+import org.testcontainers.images.RemoteDockerImage;
+
+public class ImageNameResolverProvider {
```

Comment:
This name smells, to be honest. 

This is more like an `ImageIdProvider`, with an overridable pull policy, isn't it?

Reply:
It's semantics I guess, as I see it, it is a class that provides a future that, when resolved, returns the image id. Hence, ResolverProvider.

Reply:
I still don't buy why we need this class, especially given the fact that if the delegate is set, it will simply ignore the policy.
I would just add an `imagePullPolicy` field to GenericContainer and apply it on `RemoteDockerImage`.
Or change it to:
```java
remoteDockerImage.withImagePullPolicy(imagePullPolicy).get();
```
Where `withImagePullPolicy` returns a `Future` without mutating anything in `RemoteDockerImage`

Reply:
I can't say I like any of these approaches since, outside of setDockerImageName, GenericContainer does not really know what RemoteDockerImage is - it deals with Future<String>, so in order to support withImagePull policy on the Container interface we need to be able to set it regardless of what that Future is.
What should be the behavior if GenericContainer was instantiated with a Future, and then withImagePullPolicy was called?